homeassistant/components/deluge/__init__.py

```diff
@@ -1 +1,356 @@
-"""The deluge component."""
+"""The Deluge integration."""
+from datetime import timedelta
+import logging
+
+from deluge_client import DelugeRPCClient, FailedToReconnectException
+import voluptuous as vol
+
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+
+from .const import (
+    ATTR_TORRENT,
+    DATA_UPDATED,
+    DEFAULT_NAME,
+    DEFAULT_PORT,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    SERVICE_ADD_TORRENT,
+)
+from .errors import CannotConnect, PasswordError, UserNameError, UnknownError
+
+_LOGGER = logging.getLogger(__name__)
+
+DELUGE_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Required(CONF_HOST): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+            vol.Required(CONF_USERNAME): cv.string,
+            vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+            vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+            vol.Optional(
+                CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+            ): cv.time_period,
+        }
+    )
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [DELUGE_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+PLATFORMS = ["sensor", "switch"]
+
+SERVICE_ADD_TORRENT_SCHEMA = vol.Schema(
+    {vol.Required(ATTR_TORRENT): cv.string, vol.Required(CONF_HOST): cv.string}
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Import the Deluge component from config."""
+    if DOMAIN in config:
+        for entry in config[DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": SOURCE_IMPORT}, data=entry
+                )
+            )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Deluge from a config entry."""
+    client = DelugeClient(hass, entry)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
+
+    if not await client.async_setup():
+        return False
```

Comment:
Either remove the client from `hass.data` here or only add it if set up is successful.

Reply:
Also, if it's a connection error, we should raise `ConfigEntryNotReady` as it could be that Deluge is temporarily unavailable

Reply:
- I will add it once it is successful.
- ConfigEntryNotReady is raised in async_setup(). The same has been done in other integrations like `transmission`.

Reply:
`async_setup` doesn't know about config entries, so raising there won't work.  Only inside `async_setup_entry` does it work

Reply:
It's not the component `async_setup` function, it's the client `async_setup` method. That is called inside `async_setup_entry`. So everything is good.

Reply:
Ah, my bad. Sorry.